rfcs/text/0000-multi-data-collection-dataset-processing.md

```diff
@@ -0,0 +1,120 @@
+
+This RFC defines a general mechanism for grouping bundles, here called *data groups*. By providing a general method for grouping bundles, this proposal provides a mechanism for addressing various tasks that cross bundle boundaries.  This has impact on analysis ,pipelines, algorithmic complexity of data processing, data consistency, and data set quality control.  The current bundle grouping, described in detail here, is a submission to a project.  However, the concept generalizes in a manner that other groupings can be defined to accommodate future needs.
+
+## Author(s)
+- [Nick Barkas](mailto:barkasn@broadinstitute.org)
+- [Mark Diekhans](mailto:markd@ucsc.edu)
+
+## Shepherd
+- [Nick Barkas](mailto:barkasn@broadinstitute.org)
+
+## Motivation
+The current DCP model of processing has the scope of a single data bundle, which contains a single assay or analysis.  There is no mechanism to know when a group of bundles that need to be processed together is complete and available for further processing.  This results in the limitation that data processing is linear and is entirely based on ingestion packaging resulting in multiple problems, as outlined below.
+
+#### Multi-input analysis 
+Analysis pipelines may require input from multiple different assays. To perform such analyses, they must be triggered only when all input data are available. The current mechanism of bundle notifications is not effective when data spans multiple bundles because there is no clear set of rules to identify the bundles that need to be co-processed.  This limits the downstream process to the scope of a single bundle and requires Ingest to define the scope of all subsequent processing without foreknowledge of future needs.
+
+An immediate need for analysis pipelines to process 10X V2 scRNA-seq datasets that span multiple bundles exists. The need for co-processing multiple bundles is not however limited to this scenario and extends to any other NGS modality that involves repeated sequencing of the same library. Furthermore, the need to co-process data may extend to other data modalities that the project will accept in the future.
+
+In the future, a need to run analysis processes where input sets span multiple projects may arise. A mechanism to specify processing data collections that are not restricted to a single project or submissions will be required to support this functionality.
+
+#### Algorithmic complexity and performance
+Producing combined metadata or data from multiple assay bundles using the bundle event system results in O(N^2) algorithmic complexity, where N is the number of data bundles in some grouping. As each bundle event arrives, it must be combined with accumulated metadata from previous bundles. Additionally, there is a race condition on reading and writing the accumulated results that must be handled as the order of event receipts is not guaranteed.
```

Comment:
Could you elaborate a bit more on what the race condition is here?

Reply:
Building accumulated data file in S3 from contents in multiple bundles.  For instance, creating a project metadata TSV:

- bundle event A handler reads project TSV
- bundle event B handler reads project TSV
- bundle event A handler adds A metadata to project metadata
- bundle event A handler writes project TSV
- bundle event B handler adds B metadata to project metadata
- bundle event B handler writes project TSV, overwriting A's updates

This is also O(N^2)